File: stacks/backend/CPython/chainrice_bridge/tax_api.py
"""
Tax API client for ChainRice services.
"""

from typing import Dict, Any, List, Optional
import httpx
from .exceptions import APIError


class TaxApiClient:
    """
    Client for ChainRice Tax API operations.
    """

    def __init__(
        self,
        http_client: httpx.AsyncClient,
        base_url: str,
        api_key: Optional[str] = None,
    ):
        self.http_client = http_client
        self.base_url = base_url
        self.api_key = api_key

    async def calculate_tax(self, tax_data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Calculate tax for given amount and tax type.

        Args:
            tax_data: Tax calculation data

        Returns:
            Tax calculation result
        """
        try:
            response = await self.http_client.post(
                f"{self.base_url}/api/v1/tax/calculate", json=tax_data
            )
            response.raise_for_status()
            return response.json()
        except httpx.HTTPError as e:
            raise APIError(f"Failed to calculate tax: {str(e)}")

    async def get_tax_rates(self) -> Dict[str, Any]:
        """
        Get tax rates for different categories.

        Returns:
            Tax rates data
        """
        try:
            response = await self.http_client.get(f"{self.base_url}/api/v1/tax/rates")
            response.raise_for_status()
            return response.json()
        except httpx.HTTPError as e:
            raise APIError(f"Failed to get tax rates: {str(e)}")

    async def create_contractor(
        self, contractor_data: Dict[str, Any]
    ) -> Dict[str, Any]:
        """
        Create a new contractor.

        Args:
            contractor_data: Contractor data dictionary

        Returns:
            Created contractor data
        """
        try:
            response = await self.http_client.post(
                f"{self.base_url}/api/v1/contractors", json=contractor_data
            )
            response.raise_for_status()
            return response.json()
        except httpx.HTTPError as e:
            raise APIError(f"Failed to create contractor: {str(e)}")

    async def get_contractor(self, contractor_id: str) -> Dict[str, Any]:
        """
        Get contractor by ID.

        Args:
            contractor_id: Contractor ID

        Returns:
            Contractor data
        """
        try:
            response = await self.http_client.get(
                f"{self.base_url}/api/v1/contractors/{contractor_id}"
            )
            response.raise_for_status()
            return response.json()
        except httpx.HTTPError as e:
            raise APIError(f"Failed to get contractor {contractor_id}: {str(e)}")

    async def list_contractors(
        self, page: int = 1, page_size: int = 10
    ) -> Dict[str, Any]:
        """
        List contractors with pagination.

        Args:
            page: Page number
            page_size: Number of items per page

        Returns:
            List of contractors with pagination info
        """
        try:
            response = await self.http_client.get(
                f"{self.base_url}/api/v1/contractors",
                params={"page": page, "pageSize": page_size},
            )
            response.raise_for_status()
            return response.json()
        except httpx.HTTPError as e:
            raise APIError(f"Failed to list contractors: {str(e)}")

    async def get_calculation_history(
        self, page: int = 1, page_size: int = 10
    ) -> Dict[str, Any]:
        """
        Get calculation history.

        Args:
            page: Page number
            page_size: Number of items per page

        Returns:
            Calculation history with pagination info
        """
        try:
            response = await self.http_client.get(
                f"{self.base_url}/api/v1/calculations/history",
                params={"page": page, "pageSize": page_size},
            )
            response.raise_for_status()
            return response.json()
        except httpx.HTTPError as e:
            raise APIError(f"Failed to get calculation history: {str(e)}")

    async def health_check(self) -> bool:
        """
        Health check for tax API.

        Returns:
            True if healthy
        """
        try:
            response = await self.http_client.get(f"{self.base_url}/health")
            return response.status_code == 200
        except:
            return False

File: stacks/backend/BEAM/lib/chainrice_bridge/tax_api.ex
defmodule ChainRice.Bridge.TaxApi do
  @moduledoc """
  Client for ChainRice Tax API operations.
  """

  alias ChainRice.Bridge.{Config, HttpClient}

  @doc """
  Create a new tax API client.
  """
  def new do
    %{config: Config.get()}
  end

  @doc """
  Calculate tax for given amount and tax type.
  """
  def calculate_tax(client, tax_data) do
    url = "#{client.config.tax_api_url}/api/v1/tax/calculate"
    
    case HttpClient.post(url, tax_data, headers(client)) do
      {:ok, response} -> {:ok, Jason.decode!(response.body)}
      {:error, reason} -> {:error, reason}
    end
  end

  @doc """
  Get tax rates for different categories.
  """
  def get_tax_rates(client) do
    url = "#{client.config.tax_api_url}/api/v1/tax/rates"
    
    case HttpClient.get(url, headers(client)) do
      {:ok, response} -> {:ok, Jason.decode!(response.body)}
      {:error, reason} -> {:error, reason}
    end
  end

  @doc """
  Create a new contractor.
  """
  def create_contractor(client, contractor_data) do
    url = "#{client.config.tax_api_url}/api/v1/contractors"
    
    case HttpClient.post(url, contractor_data, headers(client)) do
      {:ok, response} -> {:ok, Jason.decode!(response.body)}
      {:error, reason} -> {:error, reason}
    end
  end

  @doc """
  Get contractor by ID.
  """
  def get_contractor(client, contractor_id) do
    url = "#{client.config.tax_api_url}/api/v1/contractors/#{contractor_id}"
    
    case HttpClient.get(url, headers(client)) do
      {:ok, response} -> {:ok, Jason.decode!(response.body)}
      {:error, reason} -> {:error, reason}
    end
  end

  @doc """
  List contractors with pagination.
  """
  def list_contractors(client, opts \\ []) do
    page = Keyword.get(opts, :page, 1)
    page_size = Keyword.get(opts, :page_size, 10)
    url = "#{client.config.tax_api_url}/api/v1/contractors?page=#{page}&pageSize=#{page_size}"
    
    case HttpClient.get(url, headers(client)) do
      {:ok, response} -> {:ok, Jason.decode!(response.body)}
      {:error, reason} -> {:error, reason}
    end
  end

  @doc """
  Get calculation history.
  """
  def get_calculation_history(client, opts \\ []) do
    page = Keyword.get(opts, :page, 1)
    page_size = Keyword.get(opts, :page_size, 10)
    url = "#{client.config.tax_api_url}/api/v1/calculations/history?page=#{page}&pageSize=#{page_size}"
    
    case HttpClient.get(url, headers(client)) do
      {:ok, response} -> {:ok, Jason.decode!(response.body)}
      {:error, reason} -> {:error, reason}
    end
  end

  @doc """
  Health check for tax API.
  """
  def health_check do
    client = new()
    url = "#{client.config.tax_api_url}/health"
    
    case HttpClient.get(url, headers(client)) do
      {:ok, response} -> {:ok, response.status == 200}
      {:error, _reason} -> {:ok, false}
    end
  end

  defp headers(client) do
    base_headers = [
      {"content-type", "application/json"},
      {"accept", "application/json"}
    ]
    
    if client.config.api_key != "" do
      [{"authorization", "Bearer #{client.config.api_key}"} | base_headers]
    else
      base_headers
    end
  end
end
